### PR TITLE
La 213 permissions

### PR DESCRIPTION
**Pull request by Vadym-Lopatka**
[LA-213]

[LA-213]: https://selina.atlassian.net/browse/LA-213?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by selina-bot-user** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **vadym-lopatka**
You can assign the PR to them by writing `/assign @vadym-lopatka` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=Selina-Finance%2Fms-lending-api).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/Selina-Finance/ms-lending-api/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by selina-bot-user** [comment]
@Vadym-Lopatka: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr | 84d494d40e985b3705c1aae8611a4482596d3c5f | [link](http://dashboard-jx.selinadev.co.uk/namespaces/jx/pipelineruns/finance-ms-lending-api-pr-115-pr-5psrr) | `/test pr`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by selina-bot-user** [comment]
@Vadym-Lopatka: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr | 817f80add25a40ef88ee5b781085c39a3be8b899 | [link](http://dashboard-jx.selinadev.co.uk/namespaces/jx/pipelineruns/finance-ms-lending-api-pr-115-pr-lv8m4) | `/test pr`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: src/main/java/com/selina/lending/api/controller/CreditCommitmentsController.java [hunk 1]
```diff
@@ -46,6 +47,7 @@ public CreditCommitmentsController(UpdateCreditCommitmentsService updateCreditCo
     }
 
     @Override
+    @Permission(resource = "CC", scope = "Update")
```
Comment (bpoon-selina) [inline src/main/java/com/selina/lending/api/controller/CreditCommitmentsController.java:50]: Can we not use the full name for the resource? CC will not make much sense to someone else

File: src/main/java/com/selina/lending/internal/dto/AskedResource.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ *
+ */
+
+package com.selina.lending.internal.dto;
+
+import lombok.Builder;
+
+@Builder
+public record AskedResource(
```
Comment (bpoon-selina) [inline src/main/java/com/selina/lending/internal/dto/AskedResource.java:23]: Can we renamed to RequestedResource?

File: src/main/java/com/selina/lending/internal/service/permissions/annotation/PermissionAspect.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        this.repository = repository;
+    }
+
+    @Around("@annotation(permission)")
+    public Object permissionCheck(ProceedingJoinPoint joinPoint, Permission permission) throws Throwable {
+        var authentication = SecurityContextHolder.getContext().getAuthentication();
+        var userToken = (Jwt) authentication.getPrincipal();
+
+        var permittedResources = repository.getByUserToken(userToken.getTokenValue());
+        var askedResource = AskedResource.builder()
```
Comment (bpoon-selina) [inline src/main/java/com/selina/lending/internal/service/permissions/annotation/PermissionAspect.java:52]: use requestedResource

File: src/main/java/com/selina/lending/api/controller/CreditCommitmentsController.java (hunk 1 above)
Comment (bpoon-selina) [reply src/main/java/com/selina/lending/api/controller/CreditCommitmentsController.java:50]: I will leave this for now as I see further below we have DIP_CC etc and the resource names will be long

File: src/main/java/com/selina/lending/api/controller/DIPOperations.java [hunk 4]
```diff
@@ -136,5 +137,6 @@ ResponseEntity<ApplicationResponse> updateDipApplication(@Parameter(description
             @ApiResponse(responseCode = "403", content = @Content)
     })
     @PostMapping(value = "/dip")
+    @Permission(resource = "DIP", scope="Create")
```
Comment (bpoon-selina) [inline src/main/java/com/selina/lending/api/controller/DIPOperations.java:140]: You have this annotation in the controllers too.  Can we keep all annotations in the interface *Operations files

File: src/main/java/com/selina/lending/internal/service/permissions/PermissionServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.List;
+
+
+@Slf4j
+@Service
+public class PermissionServiceImpl implements PermissionService {
+
+    @Override
+    public boolean isAccessDenied(AskedResource asked, List<Resource> permitted) {
+        log.debug("Checking access of asked resource: {} in permitted: {}", asked, permitted);
```
Comment (bpoon-selina) [inline src/main/java/com/selina/lending/internal/service/permissions/PermissionServiceImpl.java:34]: Can we make this info logging please?

File: src/main/java/com/selina/lending/internal/service/permissions/annotation/PermissionAspect.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                .name(permission.resource())
+                .scope(permission.scope())
+                .build();
+
+        if (service.isAccessDenied(askedResource, permittedResources)) {
+            throw new AccessDeniedException("Sorry, but you have no access to this resource");
```
Comment (bpoon-selina) [inline src/main/java/com/selina/lending/internal/service/permissions/annotation/PermissionAspect.java:58]: Can we add an error log here

**Review by bpoon-selina — CHANGES_REQUESTED** [review]

File: src/main/java/com/selina/lending/internal/service/permissions/annotation/PermissionAspect.java (hunk 6 above)
Comment (Vadym-Lopatka) [reply src/main/java/com/selina/lending/internal/service/permissions/annotation/PermissionAspect.java:58]: +

File: src/main/java/com/selina/lending/internal/service/permissions/PermissionServiceImpl.java (hunk 5 above)
Comment (Vadym-Lopatka) [reply src/main/java/com/selina/lending/internal/service/permissions/PermissionServiceImpl.java:34]: sure

File: src/main/java/com/selina/lending/api/controller/DIPOperations.java (hunk 4 above)
Comment (Vadym-Lopatka) [reply src/main/java/com/selina/lending/api/controller/DIPOperations.java:140]: Thanks, forgot to remove it.

I wanted to do this as it better fits our approach but in this way, the Permission annotation does not trigger PermissionAspect.
From my googling, I understood that it is impossible to achieve such behavior when you mark the interface methods and then implement the interface.
This is why I marked the controller methods themselves.

File: src/main/java/com/selina/lending/config/security/clientOAuth2/MainAuthOauth2Configuration.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package com.selina.lending.config.security.clientOAuth2;
+
+import feign.RequestInterceptor;
+import org.springframework.context.annotation.Bean;
+
+import static org.springframework.http.HttpHeaders.AUTHORIZATION;
+
+public class MainAuthOauth2Configuration {
```
[the remark this answers is not in the document]
Comment (Vadym-Lopatka) [reply src/main/java/com/selina/lending/config/security/clientOAuth2/MainAuthOauth2Configuration.java:25]: Ok

File: src/main/java/com/selina/lending/internal/dto/AskedResource.java (hunk 2 above)
Comment (Vadym-Lopatka) [reply src/main/java/com/selina/lending/internal/dto/AskedResource.java:23]: yeap

File: src/main/java/com/selina/lending/internal/service/permissions/annotation/PermissionAspect.java (hunk 3 above)
Comment (Vadym-Lopatka) [reply src/main/java/com/selina/lending/internal/service/permissions/annotation/PermissionAspect.java:52]: ok

File: src/main/java/com/selina/lending/api/controller/ProductController.java [hunk 8]
```diff
@@ -37,6 +38,7 @@ public class ProductController implements ProductOperations{
     }
 
     @Override
+    @Permission(resource = "DIP", scope = "Update")
```
[the remark this answers is not in the document]
Comment (Vadym-Lopatka) [reply src/main/java/com/selina/lending/api/controller/ProductController.java:41]: Let's jump onto the call to discuss it

**Comment by selina-bot-user** [comment]
@Vadym-Lopatka: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr | 6f5ee10026fb37a4341a10d21446e64b268b9a11 | [link](http://dashboard-jx.selinadev.co.uk/namespaces/jx/pipelineruns/finance-ms-lending-api-pr-115-pr-d7dj9) | `/test pr`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by selina-bot-user** [comment]
@Vadym-Lopatka: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr | 6f5ee10026fb37a4341a10d21446e64b268b9a11 | [link](http://dashboard-jx.selinadev.co.uk/namespaces/jx/pipelineruns/finance-ms-lending-api-pr-115-pr-mjw72) | `/test pr`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by selina-bot-user** [comment]
:star: PR built and available in a preview **selina-finance-ms-lending-api-pr-115** [here](http://ms-lending-api-pr115.selinadev.co.uk)

**Comment by selina-bot-user** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[bpoon-selina](https://github.com/Selina-Finance/ms-lending-api/pull/115#issuecomment-1378971144 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=Selina-Finance%2Fms-lending-api).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/Selina-Finance/ms-lending-api/blob/main/OWNERS)~~ [bpoon-selina]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**